src/resolvers/api/getNode.js

```diff
@@ -0,0 +1,147 @@
+
+  const sponsors = []
+
+  // sponsor
+  const getSponsorQuery = `select sponsor_id from pki_events where node_id = '${urbitId}' order by time desc limit 1;`
+  const getSponsorResponse = await query(getSponsorQuery)
+  const getSponsorResponseRow = _get(getSponsorResponse, 'rows[0]') || []
+
+  const sponsor = getSponsorResponseRow ? _get(getSponsorResponseRow, 'sponsor_id') || null : null
+
+  sponsors.push(sponsor)
+
+  // sponsor's sponsor
+  const getSponsorsSponsorQuery = `select sponsor_id from pki_events where node_id = '${sponsor}' order by time desc limit 1;`
+  const getSponsorsSponsorResponse = await query(getSponsorsSponsorQuery)
+  const getSponsorsSponsorResponseRow = _get(getSponsorsSponsorResponse, 'rows[0]') || []
+  const sponsorsSponsor = getSponsorsSponsorResponseRow ? _get(getSponsorsSponsorResponseRow, 'sponsor_id') || null : null
+
+  if (sponsorsSponsor) {
+    sponsors.push(sponsorsSponsor)
+  }
+
+  // status
+  let statusId
+  let getStatusIdQuery = `select count(result) from radar where ship_name = '${urbitId}';`
+  let getStatusIdResponse
+  try {
+    getStatusIdResponse = await query(getStatusIdQuery)
+
+    if (_get(getStatusIdResponse, 'rows[0].count') === '0') {
+      getStatusIdQuery = `select count(*) from pki_events where node_id = '${urbitId}' and event_type_id = '6';`
+      getStatusIdResponse = await query(getStatusIdQuery)
+
+      if (_get(getStatusIdResponse, 'rows[0].count') === '0') {
+        getStatusIdQuery = `select count(*) from pki_events where node_id = '${urbitId}' and event_type_id = '7';`
+        getStatusIdResponse = await query(getStatusIdQuery)
+
+        if (_get(getStatusIdResponse, 'rows[0].count') === '0') {
+          getStatusIdQuery = `select count(*) from pki_events where node_id = '${urbitId}' and address = '0x86cd9cd0992f04231751e3761de45cecea5d1801' or address = '0x8c241098c3d3498fe1261421633fd57986d74aea';`
+          getStatusIdResponse = await query(getStatusIdQuery)
+
+          if (_get(getStatusIdResponse, 'rows[0].count') === '0') {
+            statusId = 1
+          } else {
+            statusId = 2
+          }
+        } else {
+          statusId = 3
+        }
+      } else {
+        statusId = 4
+      }
+    } else {
+      statusId = 5
+    }
+  } catch (error) {
+    console.log(`addDataResponse error: ${error}`)
+    throw error
+  }
+
+  // kids
+  const getKidsQuery = `select distinct node_id from pki_events where sponsor_id = '${urbitId}';`
+  const getKidsResponse = await query(getKidsQuery)
+  const getKidsResponseRows = _get(getKidsResponse, 'rows') || []
+  const kids = []
+
+  for (const i in getKidsResponseRows) {
+    const nodeId = _get(getKidsResponseRows, `[${i}].node_id`) || null
+    kids.push(nodeId)
+  }
+
+  // continuity number
+  const getContinuityNumberQuery = `select continuity_number from pki_events where node_id = '${urbitId}' order by time desc limit 1;`
```

Comment:
Reading from `pki_events` this way is going to be very inefficient. It means each query of a single node is going to have to result in several database queries, which won't be acceptable with respect to performance. This is why I had a `node` table in the original entity-relationship diagram. There are two approaches that can be used here:

1. After initially populating the `pki_event` table, run another job that populates the `node` table with all of these calculations precomputed. This could be done by parsing the `pki_event` data that originally went into the table, or from querying the table.
2. Create the `node` table as a materialized view of `pki_event`.

I'm generally in favor of (2), as it means we don't have to duplicate any data. It does, however, mean that you'll need to learn how to create one of these. I'd be happy to assist in showing you how to do this.

Reply:
To get an idea of the size of the problem here, this single function issues 15 different database queries. Your implementation of `getNodes` calls this function. This means that a modest query of 10 nodes would result in 150 distinct database queries.

By using either of the strategies mentioned above you could bring this number of queries down to **only one** no matter how many nodes are requested (SQL has built-in limit & offset capabilities).

Reply:
I definitely want to do option 2 as it seems the best technical approach. I've spent some time trying to cobble the materialized view together and have gotten part of it, but I do think this is something I'll need some support from you on. We can discuss on our upcoming meeting.

Reply:
I definitely want to do option 2 as it seems the best technical approach. I've spent some time trying to cobble the materialized view together and have gotten part of it, but I do think this is something I'll need some support from you on. We can discuss on our upcoming meeting.

Reply:
This is now done. Thanks for your help constructing that query!